### PR TITLE
feat(updaters): add Gradle version catalog TOML updater

### PR DESCRIPTION
**Pull request by ruromero**
## Summary
- New `src/updaters/toml_updater.js` module with `updateTomlVersions()` API
- Supports centralized `version.ref`, inline `version = "x.y.z"`, and string shorthand patterns
- Position-based regex replacement preserves TOML comments and formatting
- Handles `group`/`name` notation alongside `module` notation
- 21 new tests covering all patterns, edge cases, and error handling

Implements [TC-5412](https://redhat.atlassian.net/browse/TC-5412)

## Test plan
- [x] Centralized version updates via `version.ref`
- [x] Inline version updates
- [x] String shorthand updates
- [x] Group/name notation support
- [x] Formatting and comment preservation
- [x] Malformed TOML fail-safe
- [x] Idempotency verification
- [x] `npm run lint` passes
- [x] `npm test` passes

[TC-5412]: https://redhat.atlassian.net/browse/TC-5412?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

## Summary by Sourcery

Add a TOML updater for Gradle version catalogs and comprehensive tests and fixtures to validate version update behavior.

New Features:
- Introduce updateTomlVersions helper to update dependency versions in Gradle libs.versions.toml catalogs using centralized and inline version declarations.

Tests:
- Add extensive toml_updater test suite and TOML fixtures covering centralized refs, inline and shorthand versions, edge cases, malformed input, and idempotency.

### CODE REVIEW

File: src/updaters/toml_updater.js [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+		}
+	}
+	return undefined
+}
+
+/**
+ * Replaces a version value in the [versions] section using position-based string replacement.
+ * Targets lines matching: key = "oldVersion"
+ * @param {string} content - raw TOML content
+ * @param {string} section - section name (e.g. "versions")
```
Comment (a-oren) [inline src/updaters/toml_updater.js]: section parameter in replaceVersionInSection is declared but never used

File: src/updaters/toml_updater.js [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+ * @returns {string} updated content
+ */
+function replaceVersionInSection(content, section, key, oldVersion, newVersion) {
+	const escapedKey = escapeRegExp(key)
+	const escapedOld = escapeRegExp(oldVersion)
+	const pattern = new RegExp(
+		`^(\\s*${escapedKey}\\s*=\\s*")${escapedOld}("\\s*)$`,
+		'm'
+	)
+	return content.replace(pattern, `$1${newVersion}$2`)
```
Comment (a-oren) [inline src/updaters/toml_updater.js]: $ in newVersion corrupts the output.
use a replacer function instead of a replacement string in both replaceVersionInSection and replaceInlineVersion.

lines 245 and 253 have the same issue.

**Review by a-oren — CHANGES_REQUESTED** [review]

File: src/updaters/toml_updater.js (hunk 2 above)
Comment (ruromero) [reply src/updaters/toml_updater.js]: [sdlc-workflow/verify-pr] Classified as **code change request** — sub-task [TC-5447](https://redhat.atlassian.net/browse/TC-5447) created to address this feedback. Runtime verification confirmed: `newVersion="ver$1ify"` produces corrupted output `vermyver = "ify"`. Also confirmed `$$` and `$&` corruption.

File: src/updaters/toml_updater.js (hunk 1 above)
Comment (ruromero) [reply src/updaters/toml_updater.js]: [sdlc-workflow/verify-pr] Classified as **code change request** — sub-task [TC-5448](https://redhat.atlassian.net/browse/TC-5448) created to address this feedback.

File: src/updaters/toml_updater.js [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+function replaceInlineVersion(content, alias, oldVersion, newVersion) {
+	const escapedAlias = escapeRegExp(alias)
+	const escapedOld = escapeRegExp(oldVersion)
+	const pattern = new RegExp(
+		`^(\\s*${escapedAlias}\\s*=\\s*\\{[^}]*version\\s*=\\s*")${escapedOld}("[^}]*\\}\\s*)$`,
+		'm'
+	)
+	const result = content.replace(pattern, `$1${newVersion}$2`)
+	if (result !== content) {
+		return result
```
[the remark this answers is not in the document]
Comment (ruromero) [reply src/updaters/toml_updater.js:256]: [sdlc-workflow/verify-pr] Classified as **code change request** — sub-task [TC-5449](https://redhat.atlassian.net/browse/TC-5449) created to address this feedback. Runtime verification confirmed: caller unconditionally reports success even when replacement regex fails to match.

**Comment by ruromero** [comment]
## Verification Report for TC-5412 (commit 8edd65e)

| Check | Result | Details |
|-------|--------|---------|
| Review Feedback | WARN | 3 code change requests → 3 sub-tasks created (TC-5447, TC-5448, TC-5449) |
| Root-Cause Investigation | N/A | Defects are universal knowledge (JS regex replacement semantics); implement-task phase gap |
| Scope Containment | PASS | 3 files match task spec exactly, 0 out-of-scope, 0 unimplemented |
| Diff Size | PASS | +638 lines across 3 new files; proportionate for updater module + tests + fixture |
| Commit Traceability | PASS | Commit 8edd65ea references TC-5412 in body |
| Sensitive Patterns | PASS | No secrets or credentials detected |
| CI Status | PASS | All 5 CI checks pass (lint/test Node 22+24, Sourcery, PR title, commit messages) |
| Acceptance Criteria | WARN | 5/6 met; criterion #2 (inline versions) has 2 edge-case bugs flagged by reviewer |
| Test Quality | PASS | 21 tests with JSDoc docs, no repetitive patterns; Eval Quality: N/A |
| Test Change Classification | ADDITIVE | New test file with 21 test cases |
| Verification Commands | PASS | `npm run lint`: 0 errors; `npm test`: all 21 toml_updater tests pass |

### Overall: WARN

### Runtime Verification (manual)

The `updateTomlVersions` function was exercised directly through 17 programmatic test scenarios covering all declared patterns plus adversarial edge cases:

| # | Scenario | Result |
|---|----------|--------|
| 1–9 | Centralized ref, inline, shorthand, group/name, BOM-managed, nonexistent, multi-change, idempotency, malformed TOML | All PASS |
| 10 | Comment & formatting preservation | PASS |
| 11 | Regex-special chars in version (`1.0.0+build.123`) | PASS — `escapeRegExp()` handles correctly |
| 12 | Partial version match (`3.8.4` vs `3.8.40`) | PASS — regex anchoring prevents cross-contamination |
| 13 | Extra whitespace around `=` signs | PASS |
| 14 | Multiline library entries | PASS |
| 15 | Shared version.ref (two libraries, one update) | PASS |
| 16 | **`$1` in newVersion** | **BUG** — `ver$1ify` → `vermyver = "ify"` (regex backreference corruption) |
| 17 | Duplicate module entries | Acceptable — last-alias-wins behavior |

### Issues Requiring Attention

**1. `$` in newVersion corrupts output** ([TC-5447](https://redhat.atlassian.net/browse/TC-5447))
`replaceVersionInSection` and `replaceInlineVersion` pass `newVersion` directly into `String.prototype.replace()` replacement string. JS treats `$1`, `$&`, `$$` as special patterns. Confirmed via runtime: `newVersion='ver$1ify'` produces `vermyver = "ify"`. Fix: use a replacer function `(match, p1, p2) => ...` instead of template literal.

**2. Unused `section` parameter** ([TC-5448](https://redhat.atlassian.net/browse/TC-5448))
`replaceVersionInSection(content, section, key, oldVersion, newVersion)` declares `section` but never references it. Dead code.

**3. `replaceInlineVersion` false positives** ([TC-5449](https://redhat.atlassian.net/browse/TC-5449))
When neither regex matches, the function returns original content unchanged, but caller unconditionally pushes to `applied[]`. Should compare before/after and push to `skipped` if unchanged.

---
*This comment was AI-generated by [sdlc-workflow/verify-pr](https://github.com/RHEcosystemAppEng/sdlc-plugins) v0.13.7.*

[TC-5447]: https://redhat.atlassian.net/browse/TC-5447?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Review by a-oren — APPROVED** [review]
